[✨feat] Filter 구현

jsoonworld opened the pull request:
# 📄 Work Description  
- `Filter` 도메인 엔티티를 구현했습니다.  
- 필드는 `JobType`, `Grade`, `WorkingPeriod`, `StartDate`로 구성되며, `StartDate`는 VO로 설계되어 연월 정보의 응집도를 높였습니다.  
- `update()` 메서드를 통해 필드 전체를 변경할 수 있도록 도메인 메서드를 정의했습니다.  
- `BaseRootEntity`를 상속받아 공통 엔티티 기능도 유지했습니다.

# 💭 Thoughts  
- 단순한 필드 나열이 아닌 도메인 관점에서 의미 있는 속성들로 구성하기 위해 `StartDate`를 도입했습니다.  
- `StartDate`는 `Year`와 `Month` VO를 조합한 값 객체로, 연월에 대한 유효성 검증 책임도 함께 가집니다.  
- `Filter` 내부에서 값 변경은 `update()` 메서드를 통해 이루어지며, 변경 흐름을 명시적으로 추적 가능하게 했습니다.  
- getter 메서드는 의미 전달을 명확히 하기 위해 `jobType()`, `grade()` 등의 형태로 제공하였습니다.

# ✅ Testing Result  
![스크린샷 2025-05-18 오전 1 26 08](https://github.com/user-attachments/assets/c390bc34-6c10-46d9-84ff-0214d18d0fb5)


# 🗂 Related Issue  
- closes #34

jsoonworld committed:
feat(Month) : 월 VO 구현

1~12 범위의 월 값을 표현하는 Month VO를 정의했습니다.
유효성 검증을 포함하고, equals 및 hashCode를 오버라이드하여 값 객체로 사용할 수 있도록 구성했습니다.

jsoonworld committed:
feat(Year) : 연도 VO 구현

1900보다 큰 값을 표현하는 Year VO를 정의했습니다.
유효성 검증을 포함하고, equals 및 hashCode를 오버라이드하여 값 객체로 사용할 수 있도록 구성했습니다.

jsoonworld committed:
feat(StartDate) : 근무 시작일 VO 구현

Year와 Month를 조합하여 근무 시작일을 표현하는 StartDate VO를 구현했습니다.
JPA 임베디드 매핑을 위해 AttributeOverride를 적용하고, equals 및 hashCode를 오버라이드하여 값 객체로 사용할 수 있도록 구성했습니다.

jsoonworld committed:
test(Month) : 월 VO 생성 및 유효성 검증 테스트 추가

1부터 12 사이의 값을 정상적으로 생성하는 경우와
유효하지 않은 월(0 이하, 13 이상)에 대해 예외가 발생하는 경우를 검증하는 테스트를 작성했습니다.

jsoonworld committed:
test(StartDate) : StartDate VO 생성 테스트 추가

Year와 Month를 기반으로 StartDate 객체를 생성하고,
내부 값이 정확히 설정되는지를 검증하는 테스트를 작성했습니다.

jsoonworld committed:
test(Year) : Year VO 생성 및 유효성 검증 테스트 추가

유효한 연도로 Year 객체를 생성할 수 있는지와
1900년 이하 입력 시 예외가 발생하는 경우를 검증하는 테스트를 작성했습니다.

jsoonworld committed:
feat(Filter) : Filter 도메인 구현 및 StartDate VO 적용

jsoonworld committed:
test(Filter) : Filter 도메인 생성 및 수정 기능 테스트 추가

Filter.of()를 통해 객체 생성 시 초기 값이 정상적으로 저장되는지,
update() 메서드를 통해 필드 값이 변경되는지 검증하는 테스트를 작성했습니다.

jsoonworld committed:
refactor(StartDate) : VO 내 컬럼명을 제거하고 매핑 책임을 StartDate로 위임

- Year, Month VO 내부의 @Column name 속성을 제거하여 재사용성을 높였습니다.
- DB 컬럼명은 StartDate에서 @AttributeOverride를 통해 명시적으로 지정했습니다.
- equals 메서드를 간결한 표현식으로 개선했습니다.

jsoonworld committed:
refactor(Filter) : StartDate 임베디드 필드에 컬럼 매핑 명시

- StartDate 내부 Year, Month VO의 컬럼명 설정을 제거하고,
- Filter 엔티티에서 @AttributeOverrides를 통해 명확하게 컬럼명을 지정하여
  매핑 책임을 분리하고 재사용성을 높임.

jsoonworld committed:
merge(feat/#32 into feat/#34) : VO 구현 내용 통합

jsoonworld committed:
chore(StartDate): ktlint 포맷 적용 및 중복 equals 메서드 제거

jsoonworld committed:
merge(feat/#32 into feat/#34) : ktlint 수정 내용 반영 병합

jsoonworld committed:
chore(Filter): ktlint 포맷팅 적용 및 중복 코드 정리

- ktlint 기준에 맞게 코드 스타일 자동 포맷팅 수행
- 중복된 코드 및 불필요한 괄호 제거하여 가독성 개선
- @AttributeOverrides 내 쉼표 위치 등 문법 오류 수정

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/filter/Filter.kt:
```diff
@@ -0,0 +1,82 @@
+
+    fun startDate(): StartDate = startDate
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class != other::class) return false
+
+        other as Filter
+        return id != null && id == other.id
+    }
+
+    override fun hashCode(): Int = id?.hashCode() ?: 0
```
이 부분이 Scrap 엔티티에도 동일하게 있던 코드여서 질문합니당
JPA/Hibernate에서 안전하게 동작하도록 구현하려면 엔티티를 구현할 때 매번 작성해줘야 하는 코드인걸까요?

jsoonworld replied:
좋은 질문 감사합니다! 🙌  
말씀해주신 대로 `equals`/`hashCode`는 JPA에서 안전하게 동작하기 위해 **Entity마다 거의 필수적으로 구현해야 하는 부분**이에요.  
Hibernate나 JPA에서는 엔티티를 식별할 때 `id` 값을 기준으로 삼기 때문에,  
컬렉션(`Set`, `Map`) 사용 시나 영속성 컨텍스트에서 객체 비교 시 **id 기반 equals/hashCode**가 없으면 예기치 않은 버그가 발생할 수 있어요.

그래서 `Scrap`이나 `Filter`처럼 `id != null && id == other.id` 형태로 구현하는 게 일반적인 패턴이고,  
반복되는 부분이라 IDE 템플릿이나 공통 추상 클래스로 추출해서 관리하는 경우도 많습니다!

그리고 한 가지 덧붙이자면,  
**Value Object(VO)** 들에서는 오히려 `id`가 없기 때문에 **모든 값이 같을 때 같은 객체로 간주**하도록  
`equals`/`hashCode`를 반드시 오버라이드해줘야 해요.  
예를 들어 `Month`, `Year`, `StartDate` 같은 VO들은 내부 값 기준으로 정확하게 비교되도록 만들어주는 게 중요하죠 💡

Entity와 VO 각각의 역할과 비교 방식이 다르다는 점을 인지하고 맞춤 구현하는 게 핵심이라고 생각합니다.  
예리한 질문 감사해요! 😊

leeeyubin replied:
친절한 설명 덕분에 Entity와 VO에서 왜 해당 함수들을 오버라이드 해줘야 하는지 생각해본 것 같아요!
반복되는 부분은 말씀해 주신 것처럼 공통으로 추출해서 관리해도 좋을 것 같네요!👍

leeeyubin replied:
제가 Java와 Kotlin의 동등성과 동일성에 대해 공부하면서 알게 된 점이 하나 있는데요!

바로 Kotlin에서는 `data class`로 작성 시, equals, hashCode, toString을 자동으로 생성해 준다고 해요!
이는 [코틀린 공식 문서](https://kotlinlang.org/docs/data-classes.html)에서도 확인이 가능해요!

따라서 VO는 data class로 작성해 주고 함수는 별도로 오버라이드 하지 않는 것은 어떤가 싶은데 어떻게 생각하시나요? @jsoonworld

jsoonworld replied:
좋은 인사이트 공유 감사합니다! 😊
말씀해주신 것처럼 Kotlin의 data class는 equals/hashCode/toString을 자동으로 생성해줘서 VO 구현에 아주 적합하죠. 실제로 공식 문서에서도 VO에는 data class 사용을 권장하고 있고, 우리도 그런 방향으로 가면 유지보수도 더 수월해질 것 같아요.

다만 회의에서 이야기 나눈 대로, 현재 구현에서는 일단 기존 구조대로 반영하고, 이후 VO 설계와 관련된 학습과 리팩토링을 병행하면서 data class 적용 여부도 같이 논의해보면 좋을 것 같습니다!

이 부분은 별도 이슈로 등록해두고, 공유 감사합니다 🙌

leeeyubin commented on src/test/kotlin/com/terning/server/kotlin/domain/filter/FilterTest.kt:
```diff
@@ -0,0 +1,54 @@
+
+    @Test
+    @DisplayName("Filter의 값을 update로 변경할 수 있다")
+    fun updateFilter() {
+        // given
+        val filter = Filter.of(jobType, grade, workingPeriod, startDate)
+
+        // when
+        filter.update(newJobType, newGrade, newWorkingPeriod, newStartDate)
+
+        // then
+        assertThat(filter.jobType()).isEqualTo(newJobType)
+        assertThat(filter.grade()).isEqualTo(newGrade)
+        assertThat(filter.workingPeriod()).isEqualTo(newWorkingPeriod)
+        assertThat(filter.startDate()).isEqualTo(newStartDate)
```
어랏 지난 피알에 주석 달면 좋을 것 같다고 코드리뷰 남겼는데 여기서 반영해주셨네요!😂 좋습니다~

jsoonworld replied:
맞아요! 지난 PR에서 주셨던 피드백 덕분에 전체 테스트 스타일을 한번 정리해보게 되었어요 🙌  
여기는 비교할 값도 많고 흐름이 조금 더 복잡해서, **given–when–then 주석을 붙이면 훨씬 더 명확하게 읽힐 수 있겠다** 싶더라구요!

이번 기회에 전체 테스트 코드에 일관되게 포맷을 적용해두니까  
읽는 입장에서도 흐름을 따라가기 훨씬 쉬워진 것 같아요 😄  
좋은 피드백 덕분에 테스트 품질도 같이 올라간 느낌입니다. 감사해요!

leeeyubin reviewed (approved):
올려주신 PR 전부 확인했습니다~ 대왕 어프루브 드릴게요!🚀🚀🚀

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/filter/Filter.kt:
```diff
@@ -0,0 +1,82 @@
+    @Column(nullable = false)
+    private var jobType: JobType,
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false)
+    private var grade: Grade,
```
저희 ERD에 grade는 NULL로 나왔있는데 이 코드에 있는 `nullable = false`와는 관련이 없는 건지도 궁금합니당..!

jsoonworld replied:
꼼꼼하게 체크해주셔서 감사합니다! 🙌  
말씀해주신 대로 `grade`는 ERD 상 nullable인데, 제가 실수로 `nullable = false`로 지정했었네요 😅  
같은 이유로 `workingPeriod`도 null 가능하도록 되어 있어서 이 부분도 함께 수정했습니다.

그리고 확인하는 김에 보니 `jobType(직무유형)`도 ERD에서 누락되어 있었던 걸 발견해서  
이 부분 역시 null 가능하도록 반영해두었습니다!

놓치기 쉬운 부분까지 세심하게 짚어주셔서 덕분에 전체적인 제약 조건을 더 정확히 맞출 수 있었어요 🙏  
정말 감사드립니다!

jsoonworld committed:
refactor(Month) : 월 유효성 범위를 상수로 추출하여 가독성 및 유지보수성 개선

- 1~12의 매직 넘버를 MIN_MONTH, MAX_MONTH 상수로 추출
- 에러 메시지도 상수를 활용해 의미 전달력 향상
- 도메인 유효성 검증의 의도를 더 명확히 표현

jsoonworld committed:
refactor(test) : 도메인 테스트에 given-when-then 포맷 및 명확한 표현 적용

- 모든 도메인 테스트에 given/when/then 또는 expect 주석 포맷을 적용하여 테스트 흐름 가독성 향상
- label → color 등 변경된 도메인 필드에 맞춰 DisplayName 및 테스트 설명 수정
- 예외 검증 방식 개선 및 테스트 명세 명확화
- 테스트 구조의 일관성을 위해 필요한 경우 @Nested 및 파라미터 설명 포맷 정리

jsoonworld committed:
merge(feat/#30 → feat/#32) : 이전 브랜치 변경사항 반영을 위한 선병합

jsoonworld committed:
chore(format) : ktlintFormat 적용으로 코드 스타일 정렬

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/domain/filter/Filter.kt:
```diff
@@ -0,0 +1,82 @@
+        if (other == null || this::class != other::class) return false
+
+        other as Filter
+        return id != null && id == other.id
```
좋은 포인트 감사합니다! 🙌  
지금 `equals()`는 JPA 엔티티 기준으로 `id` 값이 있을 때만 동등하다고 판단하고,  
`null`인 경우엔 서로 다르다고 처리하고 있어요. 이 방식은 컬렉션 사용이나 영속성 관리 시 혼란을 방지하는 데 안전한 선택이에요!

Copilot에서 제안한 방식처럼 내용 기반 비교를 하려면,  
Value Object거나 명확한 비즈니스 식별자가 있는 경우에 적합한데,  
Entity에서는 `id` 기반 equals를 유지하는 게 더 일관성 있고 의도를 명확히 전달할 수 있어서 현재 구현이 적절하다고 생각합니다 😊
피드백 감사합니다~!

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/domain/filter/Filter.kt:
```diff
@@ -0,0 +1,82 @@
+                startDate = startDate,
+            )
+    }
+
```
좋은 피드백 감사합니다! 🙌  
말씀해주신 대로 문서화를 통해 파라미터 의도를 명시해주는 것도 중요한 접근인 것 같아요.  
다만 이 메서드의 경우, 네이밍 자체로 이미 역할과 목적이 명확하게 드러난다고 판단해서  
불필요한 문서보다는 **간결한 네이밍을 유지하는 쪽이 더 가독성과 유지보수에 유리**하다고 생각했습니다.

추후 메서드가 더 복잡해지거나 문맥이 흐려질 경우,  
그때는 KDoc을 보완적으로 도입하는 방향도 고려해보겠습니다! 😊

jsoonworld committed:
merge(feat/#32 → feat/#34) : 이전 브랜치 변경사항 반영을 위한 선병합

jsoonworld committed:
refactor(Filter) : jobType, grade, workingPeriod의 nullable 설정 제거

- 불필요한 nullable = false 제거하여 ERD 기준과 일치하도록 수정
- JPA 기본값인 nullable = true로 매핑되도록 설정 정비
